polynote-kernel/src/main/scala/polynote/kernel/context/GlobalInfo.scala

```diff
@@ -1,15 +1,82 @@
-package polynote.kernel.util
+package polynote.kernel.context
 
 import java.io.File
 import java.net.URL
 
-import scala.reflect.internal.util.{AbstractFileClassLoader, BatchSourceFile}
+import cats.effect.IO
+import polynote.kernel.lang.LanguageKernel
+import polynote.kernel.util._
+
 import scala.reflect.internal.util.ScalaClassLoader.URLClassLoader
+import scala.reflect.internal.util.{AbstractFileClassLoader, BatchSourceFile}
 import scala.reflect.io.{AbstractFile, VirtualDirectory}
 import scala.tools.nsc.Settings
 import scala.tools.nsc.interactive.Global
 
-final case class GlobalInfo(global: Global, classPath: List[File], classLoader: AbstractFileClassLoader)
+final case class GlobalInfo(global: Global, classPath: List[File], classLoader: AbstractFileClassLoader) {
+
+  type GlobalProxy = global.type
+  import global.{Symbol, TermName, Type}
+
+  private val runtimeMirror = scala.reflect.runtime.universe.runtimeMirror(classLoader)
+  private val importFromRuntime = global.internal.createImporter(scala.reflect.runtime.universe)
+
+  private def typeOf(value: Any, staticType: Option[Type]): Type = staticType.getOrElse {
+    try {
+      importFromRuntime.importType {
+        runtimeMirror.reflect(value).symbol.asType.toType
+      }
+    } catch {
+      case err: Throwable => global.NoType
+    }
+  }
+
+  def formatType(typ: global.Type): String = typ match {
+    case mt @ global.MethodType(params: List[Symbol], result: Type) =>
+      val paramStr = params.map {
+        sym => s"${sym.nameString}: ${formatType(sym.typeSignatureIn(mt))}"
+      }.mkString(", ")
+      val resultType = formatType(result)
+      s"($paramStr) => $resultType"
+    case global.NoType => "<Unknown>"
+    case _ =>
+      val typName = typ.typeSymbolDirect.name
+      val typNameStr = typ.typeSymbolDirect.nameString
+      typ.typeArgs.map(formatType) match {
+        case Nil => typNameStr
+        case a if typNameStr == "<byname>" => s"=> $a"
+        case a :: b :: Nil if typName.isOperatorName => s"$a $typNameStr $b"
+        case a :: b :: Nil if typ.typeSymbol.owner.nameString == "scala" && (typNameStr == "Function1") =>
+          s"$a => $b"
+        case args if typ.typeSymbol.owner.nameString == "scala" && (typNameStr startsWith "Function") =>
+          s"(${args.dropRight(1).mkString(",")}) => ${args.last}"
+        case args => s"$typName[${args.mkString(", ")}]"
+      }
+  }
+
+  sealed case class RuntimeValue(
+    name: GlobalProxy#TermName,
```

Comment:
Do you need `GlobalProxy`? Since you imported `TermName` from `Global`, it should work on its own, no?